src/datatablerow/utils.jl

```diff
@@ -0,0 +1,191 @@
+# Rows grouping.
+# Maps row contents to the indices of all the equal rows.
+# Used by groupby(), join(), nonunique()
+immutable RowGroupDict{T<:AbstractDataTable}
+    "source data table"
+    dt::T
+    "number of groups"
+    ngroups::Int
+    "row hashes"
+    rhashes::Vector{UInt}
+    "hashindex -> index of group-representative row"
+    gslots::Vector{Int}
+    "group index for each row"
+    groups::Vector{Int}
+    "permutation of row indices that sorts them by groups"
+    rperm::Vector{Int}
+    "starts of ranges in rperm for each group"
+    starts::Vector{Int}
+    "stops of ranges in rperm for each group"
+    stops::Vector{Int}
+end
+
+# "kernel" functions for hashrows()
+# adjust row hashes by the hashes of column elements
+function hashrows_col!(h::Vector{UInt}, v::AbstractVector)
+    @inbounds for i in eachindex(h)
+        h[i] = hash(v[i], h[i])
+    end
+    h
+end
+
+function hashrows_col!{T<:Nullable}(h::Vector{UInt}, v::AbstractVector{T})
+    @inbounds for i in eachindex(h)
+        h[i] = isnull(v[i]) ?
+               hash(Base.nullablehash_seed, h[i]) :
+               hash(unsafe_get(v[i]), h[i])
+    end
+    h
+end
+
+# should give the same hash as AbstractVector{T}
+function hashrows_col!{T}(h::Vector{UInt}, v::AbstractCategoricalVector{T})
+    # TODO is it possible to optimize by hashing the pool values once?
+    @inbounds for (i, ref) in enumerate(v.refs)
+        h[i] = hash(CategoricalArrays.index(v.pool)[ref], h[i])
+    end
+    h
+end
+
+# should give the same hash as AbstractNullableVector{T}
+function hashrows_col!{T}(h::Vector{UInt}, v::AbstractNullableCategoricalVector{T})
+    # TODO is it possible to optimize by hashing the pool values once?
+    @inbounds for (i, ref) in enumerate(v.refs)
+        h[i] = ref == 0 ?
+               hash(Base.nullablehash_seed, h[i]) :
+               hash(CategoricalArrays.index(v.pool)[ref], h[i])
+    end
+    h
+end
+
+# Calculate hash for each row
+# in an efficient column-wise manner
```

Comment:
Thanks. Indeed, better put the comment about memory access before `hashrows_col!` since this function calls it anyway, no need to repeat it.

I'd also use a standard comment rather than a docstring since currently no internal function uses docstrings.